### ISSUE TITLE
Commands - Implementing getTenants() Correctly

### ISSUE THREAD

**Issue by watercayman** [stancl/tenancy#371]
The `TenantAwareCommand` trait is a great time saver, allowing us to avoid writing loops on potentially many nightly commands.

If I add the `TenantAwareCommand` and then the `HasATenantsOption`, the docs state that if I use these traits, I won't have to change a thing in my commands.  However, when I combine these two traits, I see the following error on all commands:

> The "tenants" option does not exist.

If I remove the `HasATenantsOption` and implement `getTenants()` per the docs, it works perfectly with this code in my commands:

`protected function getTenants(): array
    {
        return tenancy()->all()->toArray();
    }`

So, I have two questions:

1. Why is the trait failing (what have I done wrong)?
2. If I know I need to process every tenant for the command, why do I need to have the extra `all()` and `$this->option('tenants')` in the `getTenants()` method?  I don't understand what this is doing / what its purpose is in this instance?

**Comment by stancl** [comment]
Hi,

Can you show me your command's constructor & handle method? Maybe that is causing issues. The trait overrides the `getOptions()` method to add an optional `--tenants` option.

https://github.com/stancl/tenancy/blob/03e21621a06bffe49dc70f1f8e081999bab68f90/src/Traits/HasATenantsOption.php#L11-L16

Then it passes the array to `all()`:

https://github.com/stancl/tenancy/blob/03e21621a06bffe49dc70f1f8e081999bab68f90/src/Traits/HasATenantsOption.php#L18-L21

The point of that is to let you specify which tenants to run the command against, but default to all tenants when you don't pass any `--tenants` arguments.

And yes, your own implementation of `getTenants()` is correct, if you don't need to be able to execute the command only for some tenants.

**Comment by watercayman** [comment]
OK - perfect - good to know the `getTenants()` is correct as is, thank you.

As to the commands, I have several, and they all failed with same message.  None of them have anything but the standard call to parent in the constructor.  Here is the constructor for all of them:

     public function __construct()
     {
         parent::__construct();
     }
Here is one `handle()`:

    public function handle()
    {
        $noticesToDelete = Notice::whereDate('expires', '<', \Carbon\Carbon::now())->pluck('id');

        Notice::destroy($noticesToDelete);

        return true;
    }

A different example:
      `public function handle(){

        $path = storage_path().'/pdfs/';
        if(\File::isDirectory($path)) {
            $files = array_diff(scandir($path), array('.', '..'));  // Remove .. and . from scandir results
            foreach ($files as $file) {
                if (file_exists($path . $file)) {
                    if ($file !== '.gitignore')
                        unlink($path . $file);
                }
            }
         }
     }
```

**Comment by stancl** [comment]
I see. Apparently this causes issues when the constructor calls the parent constructor but not `$this->specifyParameters();`, since the trait's `__construct` method is not technically parent, as it's a trait.

https://github.com/stancl/tenancy/blob/03e21621a06bffe49dc70f1f8e081999bab68f90/src/Traits/HasATenantsOption.php#L23-L28

**Comment by watercayman** [comment]
Interesting.  I can confirm that all of my commands work with the `HasATenantsOption` if I add `$this->specifyParameters(); ` to the constructor.

**Comment by stancl** [comment]
It's just that you have a constructor that calls base command `__construct()` but not the trait's `__construct()`, because a trait is not a parent. If you get rid of those empty constructors, it will default to the trait's command which will make it work too.

**Comment by watercayman** [comment]
Yes, this makes perfect sense.  And solves the issue - thank you again!

You may want to make note of what you've told me in the docs.  When running `php artisan make:command`, it includes the constructor and defaults to the constructor without the specify method:

    public function __construct() 
     { 
         parent::__construct(); 
    }

**Comment by stancl** [comment]
Good point. Adding the `documentation` label.
